File: ForecastingAirlines.R
library(forecast) #for forecasting
library(fpp)     
library(smooth)   #for smoothing techniques
library(readxl)
airline<-read_xlsx(file.choose())
View(airline)

library(tseries)  #for converting the data to time series
amts<-ts(airline$Passengers,frequency = 12,start=c(1995))
View(amts)
plot(amts)

train<-amts[1:84]
test<-amts[85:96]
train<-ts(train,frequency = 12)
test<-ts(test,frequency = 12)

hw_abg<-HoltWinters(train,alpha = 0.2,beta = 0.15,gamma = 0.05) #HoltWinters with specified alpha, beta and gamma values
hw_abg
hwabg_pred<-data.frame(predict(hw_abg,n.ahead = 12))
plot(forecast(hw_abg,h=12))
hwabg_mape<-MAPE(hwabg_pred$fit,test)*100

hw_nabg<-HoltWinters(train) #HoltWinters with optimal alpha, beta and gamma values
hw_nabg
hwnabg_pred<-data.frame(predict(hw_nabg,n.ahead =12))
plot(forecast(hw_nabg,h=12))
hwnabg_mape<-MAPE(hwnabg_pred$fit,test)*100

df_mape<-data.frame(c("hwabg_mape","hwnabg_mape"),c(hwabg_mape,hwnabg_mape))
colnames(df_mape)<-c("MAPE","VALUES")
View(df_mape)

new_model <- HoltWinters(amts) #HoltWinters with optimal alpha, beta and gamma values has the least mean absolute percentage error
forecast_new <- data.frame(predict(new_model,n.ahead=12))
plot(forecast(new_model,n.ahead=12))

########## Model Based Forecasting ###########
X<- data.frame(outer(rep(month.abb,length = 96), month.abb,"==") + 0 )# Creating dummies for 12 months
View(X)
colnames(X)<-month.abb # Assigning month names 
View(X)
airline1<-cbind(airline,X)

airline1["t"]<- 1:96
airline1["t_square"]<-airline1["t"]*airline1["t"]
attach(airline1)

train<-airline1[1:84,-c(1)]
test<-airline1[85:96,-c(1)]

linear_model<-lm(Passengers~t,data=train) #linear model
summary(linear_model)
linear_pred<-data.frame(predict(linear_model,interval='predict',newdata =test))
rmse_linear<-sqrt(mean((test$Passengers-linear_pred$fit)^2,na.rm = T))
rmse_linear

Quad_model<-lm(Passengers~t+t_square,data=train) #quadratic model
summary(Quad_model)
Quad_pred<-data.frame(predict(Quad_model,interval='predict',newdata=test))
rmse_Quad<-sqrt(mean((test$Passengers-Quad_pred$fit)^2,na.rm=T))
rmse_Quad

Add_sea_Linear_model<-lm(Passengers~t+Jan+Feb+Mar+Apr+May+Jun+Jul+Aug+Sep+Oct+Nov,data=train) # linear model with additive seasonality
summary(Add_sea_Linear_model)
Add_sea_Linear_pred<-data.frame(predict(Add_sea_Linear_model,interval='predict',newdata=test))
rmse_Add_sea_Linear<-sqrt(mean((test$Passengers-Add_sea_Linear_pred$fit)^2,na.rm=T))
rmse_Add_sea_Linear

Add_sea_Quad_model<-lm(Passengers~t+t_square+Jan+Feb+Mar+Apr+May+Jun+Jul+Aug+Sep+Oct+Nov,data=train) # quadratic model with additive seasonality
summary(Add_sea_Quad_model)
Add_sea_Quad_pred<-data.frame(predict(Add_sea_Quad_model,interval='predict',newdata=test))
rmse_Add_sea_Quad<-sqrt(mean((test$Passengers-Add_sea_Quad_pred$fit)^2,na.rm=T))
rmse_Add_sea_Quad

table_rmse<-data.frame(c("rmse_linear","rmse_Quad","rmse_Add_sea_Linear","rmse_Add_sea_Quad"),c(rmse_linear,rmse_Quad,rmse_Add_sea_Linear,rmse_Add_sea_Quad))
colnames(table_rmse)<-c("model","RMSE")
View(table_rmse)

new_model <- lm(Passengers~t+t_square+Jan+Feb+Mar+Apr+May+Jun+Jul+Aug+Sep+Oct+Nov,data=airline1)
pred_new<-data.frame(predict(new_model,newdata=airline1,interval = 'predict'))
View(pred_new)
plot(pred_new$fit,type = "o")

File: ForecastingAirlines.py
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from statsmodels.tsa.seasonal import seasonal_decompose
from statsmodels.tsa.holtwinters import SimpleExpSmoothing
from statsmodels.tsa.holtwinters import Holt
from statsmodels.tsa.holtwinters import ExponentialSmoothing
import statsmodels.graphics.tsaplots as tsa_plots

airlineData = pd.read_excel("C:/My Files/Excelr/17 - Forecasting/Assignment/Airlines+Data.xlsx")
airlineData.shape
airlineData.columns
airlineData.Month[0]
airlineData.Passengers.plot()

#################### MODEL BASED APPROACH ###################

pd.to_datetime(airlineData.Month[0],format="%m")
airlineData.Month[0].strftime("%b")
airlineData['month_name']= 0

#adding new column with Month Name for converting to dummy variables
for i in range(0,96):
    p = airlineData.Month[i]
    airlineData['month_name'][i]= airlineData.Month[i].strftime("%b")    

month_dummies = pd.get_dummies(airlineData['month_name'])
airlineData_dumm = pd.concat([airlineData,month_dummies],axis=1)

airlineData_dumm['t'] = np.arange(1,97)
airlineData_dumm["t_squared"] = airlineData_dumm["t"]*airlineData_dumm["t"]
airlineData_dumm["log_Passenger"] = np.log(airlineData_dumm["Passengers"])
airlineData_dumm = airlineData_dumm.drop('Month',axis=1)
airlineData_dumm.shape
airlineData_dumm.columns

Train = airlineData_dumm.head(84)
Test = airlineData_dumm.tail(12)

import statsmodels.formula.api as smf
linear_model = smf.ols('Passengers ~ t',data = Train).fit() #Linear
linear_model.summary()
pred_linear =  pd.Series(linear_model.predict(pd.DataFrame(Test['t'])))
rmse_linear = np.sqrt(np.mean((np.array(Test['Passengers'])-np.array(pred_linear))**2))
rmse_linear

Exp = smf.ols('log_Passenger~t',data=Train).fit() #Exponential
Exp.summary()
pred_Exp = pd.Series(Exp.predict(pd.DataFrame(Test['t'])))
rmse_Exp = np.sqrt(np.mean((np.array(Test['Passengers'])-np.array(np.exp(pred_Exp)))**2))
rmse_Exp

Quad = smf.ols('Passengers~t+t_squared',data=Train).fit() #Quadratic
pred_Quad = pd.Series(Quad.predict(Test[["t","t_squared"]]))
rmse_Quad = np.sqrt(np.mean((np.array(Test['Passengers'])-np.array(pred_Quad))**2))
rmse_Quad

add_sea = smf.ols('Passengers~Jan+Feb+Mar+Apr+May+Jun+Jul+Aug+Sep+Oct+Nov',data=Train).fit() #Additive Seasonality
pred_add_sea = pd.Series(add_sea.predict(Test[['Jan','Feb','Mar','Apr','May','Jun','Jul','Aug','Sep','Oct','Nov']]))
rmse_add_sea = np.sqrt(np.mean((np.array(Test['Passengers'])-np.array(pred_add_sea))**2))
rmse_add_sea

add_sea_Quad = smf.ols('Passengers~t+t_squared+Jan+Feb+Mar+Apr+May+Jun+Jul+Aug+Sep+Oct+Nov',data=Train).fit() #Quadratic with addititve seasonality
pred_add_sea_quad = pd.Series(add_sea_Quad.predict(Test[['Jan','Feb','Mar','Apr','May','Jun','Jul','Aug','Sep','Oct','Nov','t','t_squared']]))
rmse_add_sea_quad = np.sqrt(np.mean((np.array(Test['Passengers'])-np.array(pred_add_sea_quad))**2))
rmse_add_sea_quad

Mul_sea = smf.ols('log_Passenger~Jan+Feb+Mar+Apr+May+Jun+Jul+Aug+Sep+Oct+Nov',data = Train).fit() #Multiplicative Seasonality
pred_Mult_sea = pd.Series(Mul_sea.predict(Test))
rmse_Mult_sea = np.sqrt(np.mean((np.array(Test['Passengers'])-np.array(np.exp(pred_Mult_sea)))**2))
rmse_Mult_sea

Mul_Add_sea = smf.ols('log_Passenger~t+Jan+Feb+Mar+Apr+May+Jun+Jul+Aug+Sep+Oct+Nov',data = Train).fit() #Multiplicative Additive Seasonality
pred_Mult_add_sea = pd.Series(Mul_Add_sea.predict(Test))
rmse_Mult_add_sea = np.sqrt(np.mean((np.array(Test['Passengers'])-np.array(np.exp(pred_Mult_add_sea)))**2))
rmse_Mult_add_sea 

Mul_Add_qa_sea = smf.ols('log_Passenger~t+t_squared+Jan+Feb+Mar+Apr+May+Jun+Jul+Aug+Sep+Oct+Nov',data = Train).fit() #Multiplicative Additive Quadratic Seasonality
pred_Mult_add_qa_sea = pd.Series(Mul_Add_sea.predict(Test))
rmse_Mult_add_qa_sea = np.sqrt(np.mean((np.array(Test['Passengers'])-np.array(np.exp(pred_Mult_add_qa_sea)))**2))
rmse_Mult_add_qa_sea 

data = {"MODEL":pd.Series(["rmse_linear","rmse_Exp","rmse_Quad","rmse_add_sea","rmse_add_sea_quad","rmse_Mult_sea","rmse_Mult_add_sea","Mul_Add_qa_sea"]),"RMSE_Values":pd.Series([rmse_linear,rmse_Exp,rmse_Quad,rmse_add_sea,rmse_add_sea_quad,rmse_Mult_sea,rmse_Mult_add_sea,rmse_Mult_add_qa_sea])}
table_rmse=pd.DataFrame(data)
table_rmse

plt.plot(data['MODEL'],data['RMSE_Values'])
plt.scatter(data['MODEL'],data['RMSE_Values'])


#################### DATA DRIVEN APPROACH ######################

#Create new columns with month, year for better understaning of data
airlineData['month_name'] = airlineData.Month.dt.strftime("%b")
airlineData['year_name'] = airlineData.Month.dt.strftime("%Y")

heatmap_y_month = pd.pivot_table(data=airlineData,values="Passengers",index="year_name",columns="month_name",aggfunc="mean",fill_value=0)
sns.heatmap(heatmap_y_month,annot=True,fmt="g")

sns.boxplot(x="month_name",y="Passengers",data=airlineData)
sns.boxplot(x="year_name",y="Passengers",data=airlineData)
sns.lineplot(x="year_name",y="Passengers",hue="month_name",data=airlineData)

airlineData.Passengers.plot(label="Passenger")

for i in range(2,24,6):
    airlineData["Passengers"].rolling(i).mean().plot(label=str(i))
    
plt.legend(loc=3)

airlineData.index = pd.to_datetime(airlineData.Month,format="%b-%y")
decompose_ts_add = seasonal_decompose(airlineData.Passengers,model="additive")
decompose_ts_add.plot()
decompose_ts_mul = seasonal_decompose(airlineData.Passengers,model="multiplicative")
decompose_ts_mul.plot()

# ACF plots and PACF plots on Original data sets 
tsa_plots.plot_acf(airlineData.Passengers,lags=10)
tsa_plots.plot_pacf(airlineData.Passengers)

Train = airlineData.head(84)
Test = airlineData.tail(12)

def MAPE(pred,org):
    temp = np.abs((pred-org))*100/org
    return np.mean(temp)

# Simple Exponential Method
ses_model = SimpleExpSmoothing(Train["Passengers"]).fit()
pred_ses = ses_model.predict(start = Test.index[0],end = Test.index[-1])
MAPE(pred_ses,Test.Passengers)

# Holt method 
hw_model = Holt(Train["Passengers"]).fit()
pred_hw = hw_model.predict(start = Test.index[0],end = Test.index[-1])
MAPE(pred_hw,Test.Passengers)
tendArr = ["add", "mul", "additive", "multiplicative"]
sesonalArr = ["add", "mul", "additive", "multiplicative"]

mapeArr = [];
for i in tendArr:
    for j in sesonalArr:
        hwe_model_add_add = ExponentialSmoothing(Train["Passengers"],seasonal=j,trend=i,seasonal_periods=12,damped=True).fit()
        pred_hwe_add_add = hwe_model_add_add.predict(start = Test.index[0],end = Test.index[-1])
        mapeArr.append([i,j,MAPE(pred_hwe_add_add,Test.Passengers)])
